### ISSUE TITLE
ToggleButtonBar Selection Issue

### ISSUE THREAD

**Issue by Bilbosax** [apache/royale-asjs#1074]
Hi.  I'm trying to use the new ToggleButtonBar in my app, but I'm not sure that it is performing the way a togglebuttonbar is intended to.  In Flex, only one button could be selected at a time.  So if you selected a button, it deselected all the other buttons.  I can't seem to do this with the Royale version.  Your version allows you to have unlimited buttons in the selected state. I can get the selectedIndex of the button clicked, but I can't PURPOSELY deselect the other buttons on a button click.  Is this a bug, or am I missing how to deselect buttons?

**Comment by carlosrovira** [comment]
Hi Bilbo, 

It was designed that way, so is not a current bug, it's a feature (hehe, I always wanted to say that ;)).

Being said this, I understand your need and I think we should do that way.

**Comment by carlosrovira** [comment]
Hi, I added `allowMultipleSelection` (defaults to false, since seems the most normal use case). So to get the previous behavior people need to put it to true:

```xml
<j:ToggleButtonBar localId="tbb2" allowMultipleSelection="true" ...
```

**Comment by Bilbosax** [comment]
Great idea on the allowMultipleSelection parameter, but I don't believe it is working as you intended yet.  The first click of a button does indeed deselect the other button, but after that it all fails.  Selecting the second button, does not deselect the first button, and once they are both selected, neither one can be deselected.  They are permanently selected.

**Comment by Bilbosax** [comment]
I'm not sure that I would close this one yet.  It is almost there.  If you click a button with your new allowMultipleSelection=false parameter, it does indeed deselect the other buttons.  But if you click an already selected button, it deselects it too so that none of the buttons are selected.  I think that in the case where allowMultipleSelection is set to false, at least one button needs to be selected at all times.  Is there a way to make this happen?

**Comment by carlosrovira** [comment]
Hi Bilbo, what was the behavior of Flex regarding that? I think we can go that way or maybe add some property, but don't have clear if we should remove the actual behavior. I mean that I created the ToggleButtonBar thinking on a set of ToggleButtons and probably not always others will want to have a button selected. Other times will be the case. So probably we need to support both. But don't have right now a clear vision of what to do. We need to find a clear path on how to solve this, but is clear we need to do something.

**Comment by Bilbosax** [comment]
Hi Carlos.  So I did a little research and this is what I found.  At the link below for the Flex MX ToggleButtonBar component, the first lines state

"Only one button in the ToggleButtonBar control can be in the selected state. This means that when a user selects a button in a ToggleButtonBar control, the button stays in the selected state until the user selects a different button."

The original default is that only one button can be selected at a time, and one button must ALWAYS be selected.  They did add a _toggleOnClick_ parameter that can be set to true or false that allows a user to deselect a button that was previously selected so that no buttons are selected in the togglebuttonbar, but the default is that a button must always be selected.  Here is a link with the detailed description.

https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/mx/controls/ToggleButtonBar.html

**Comment by carlosrovira** [comment]
Hi Bilbo, please check the new changes, I think will make your concrete case work (just select one button at a time). I added "toggleOnClick", but still need to implement "-1" when current button is deselected.

I removed "allowMultipleSelection" since I think is not ok after the latest investigation. In Jewel there's still no multiple selection on list (although it should be somewhat easy to bring from Basic), and probably will be better to add at that time.

**Comment by Bilbosax** [comment]
Hi Carlos.  I tested the new ToggleButtonBar and see that you added "toggleOnClick" and got rid of "allowMultipleSelection", and it works in the sense that only one button can be selected at a time which is great.  But if I click a button that is already selected, it gets deselected, so I end up with a togglebutton bar with no selections at all.  Shouldn't the default be that one button must ALWAYS remain selected?

**Comment by carlosrovira** [comment]
Hi, 

`toggleOnClick` is false by default and I'm checking that is working ok. That's the behaviour you want. If you make it `true` then is behaving as you said. I'm still need to support selectedIndex = -1, and will work on it as soon as I can.

IOW something like this in TDJ is what you want:

```xml
<j:ToggleButtonBar localId="tbb" className="iconButtonBarFontAwesomeIcon"
					change="tbbChangeHandler(event)"
					selectedIndex="2"
					emphasis="primary" width="100%" 
					dataProvider="{listModel.iconButtonDataFW}"/>
```

**Comment by Bilbosax** [comment]
Interesting.  I am tracing out that tbb.selectedIndex is correct for every button click.  But I am not seeing it VISUALLY.

So, with toggleOnClick set to false, my app opens like this with selectedIndex=0 :

![Screenshot (71)](https://user-images.githubusercontent.com/23325601/107394579-08e38a80-6aca-11eb-8dd3-72d169385c6e.png)

Clicking the rental button changes the toggleButtonBar correctly and traces selectedIndex = 1 👍 

![Screenshot (72)](https://user-images.githubusercontent.com/23325601/107394806-3defdd00-6aca-11eb-93b5-cab7cd339d71.png)

Clicking Rental AGAIN traces out selectedIndex=1, which is correct, but visually it deselects the button 👍 

![Screenshot (73)](https://user-images.githubusercontent.com/23325601/107394961-65df4080-6aca-11eb-8415-fd487ed0abf0.png)

Are you seeing something different?  Is there a way to keep it in the selected state?

**Comment by carlosrovira** [comment]
First case is right. By default, if you don't set any other selection, selectedIndex will be 0.
Second is right too. Selecting second button will make selectedInex = 1
But something is wrong with third case. I think you are missing new CSS (the same as some days ago) since you should not be able to unselect a selected button (unless you have toggleOnClick=true).
In that config all is ok from my side.
The problem is if I set "toggleOnClick=true", then I can unselect a selected button, and as Flex docs says, it must set the selectedIndex = -1. That's the thing I need to implement.

So in short,  I think you are missing this CSS chunk:

```sass
.jewel.buttonbar
    &.toggle-on-click
        .jewel.togglebutton
            &.selected
                pointer-events: none
```

**Comment by Bilbosax** [comment]
Thanks Carlos.  I really appreciate all of your help, I am not there yet, but I am beginning to feel like Royale may actually be able to handle my needs on the web.  After a couple of your recent comments on several of my questions, I think I am having a problem with CSS on my system and I have no idea why.  I will outline what I am doing with CSS, and maybe you can tell me if it sounds like I am missing anything in my setup.

1)  I download the nightly 0.9.8 Royale SDK using the built in features of Moonshine, not using npm.  Hopefully Moonshine does a thorough job in its updates.

2) I design using VS Code.  In my asconfig file, I include the following line to provide the theme for my app 
"theme": "${royalelib}/themes/Jewel-Light-Flat-Primary-Sapphire-Theme/src/main/resources/defaults.css"

3) I use the **_SimpleCSSValuesImpl_** bead at the Application level to provide some CSS features to my application.

This is my basic setup for updating my system and applying CSS in Royale.  Yet I can't get visibility to happen correctly in MXML with the HSlider (from another post), and I can't get the ToggleButtonBar buttons to display properly as you claim they should.  I know the SDK must be getting updated because I have access to the new properties that you create, but the visual displays are not updating as you say they should.  I checked the CSS file that is generated by debugging, and the .jewel.buttonbar rules that you show above are nowhere to be found.  Why would they be missing? Is there some kind of default CSS file that is not part of the theme that I have to make sure is included in my app?  Any thoughts at all on making sure CSS on my system behaves the same on my system as it does on yours?

**Comment by carlosrovira** [comment]
Hi Bilbo, 

yeah, it seems there's some issue in your CSS. Lets see if we find the issue.

I can confirm that all is ok at framework level. For what you say, it seems you already have the new bits since the piece of code is the generated CSS from the SASS I posted earlier, so you have it too. So this makes me think that there's something on your side that is causing the problem. Could be a) a browser cache problem or b) a config in compilation problem.

a) is easy to check by clearing your browser's cache and reloading to see if all is going ok now. Or trying other browser where you don't test it already.

b) SimpleCSSValuesImpl is ok (You can see that's what we use in TDJ). The CSS changed for HSlider and for ToggleButtonBar are not in the theme since are considered "structure", not something "themeable". Just notice that in Jewel we separate CSS in "structure" and in "theme" so we have in theme projects the things we really can change for a different look and feel like colors, borders, backgrounds, and more..... In resume, both latest changes are in Jewel css. Since you are building and seeing the changes in you debug's "App.css" final css, all seems to be ok in your build. You as well will be seeing other styles from the theme you choose (flat-primary-shappire) inside App.css. So all seems to be ok from the info you provided

Side Note: if you want to change something from the theme you add a new CSS at your project level and can "override" styles. The Royale compiler will compile all css but will choose yours over the framework's lines and compose a final App.css, so App defined css is always preferred over the same css rule in framework libs.

So, from the actual info, I think is a cache problem. If not please let me know a will continue dicarding until we reach the problem.

About building: I'm using Maven build since I prefer depend on a build system instead of any IDE to ensure my build is consistent from my desktop development environment to any CI server out there. So I don't use VSCode to build, just use it for code intelligence and to catch errors when coding, but that should not be a problem. Also I remember Josh added the option to add more than one CSS as an Array, just say this so you can check that line and try different things. In Jewel you usually add 3 themes one for primary, one for secondary and one for emphasized. All of this shouldn't have to be anything with your problem, but just for your info.

Let's see if we find the issue...

**Comment by Bilbosax** [comment]
Still trying to get to the bottom of this.  I am confident that it is not a cache issue because I opened the index.html file in four different browsers and all four performed the same.  I created a simple "Hello World" app and put a ToggleBarButton in it and it performed as expected, and I placed one inside of TDJ, and it worked as expected.  So something small, somewhere, is happening in my app and I am probably going to have to build it component by component from scratch until I find it, but I have NO IDEA what it could be because I am simply not that clever.

I will say that I have noticed another small problem with ToggleButtonBar that may have gone unnoticed.  When I placed one in TDJ and in my simple "Hello World" app, it does not have a default "selected" state, and I could not create one.  So on startup, no button is selected in the ToggleBarButton, and I can't do something like toggleButtonBar.selectedIndex = 1 to force it to select a button.  Only a click on a button whose selectedIndex was not zero worked.  I could not click the first button and get it to select.  Once the second button was clicked, then everything worked as expected.  The ToggleBarButton should default to a selectedIndex of 0 unless otherwise specified, or should be able to be selected by setting the selectedIndex to any accetable value.

Now to figure out what I am doing wrong on my end...

**Comment by Bilbosax** [comment]
Carlos, I could really use another pair of eyes.  I cut out thousands of lines of code, section by section.  I have trimmed it down to a very very simple app, laid out just like TDJ with an App.mxml, a MainContent.mxml, and a single component called spinner.mxml.  Spinner.mxml contains a single ToggleButtonBar.  The entire app is less than 75 lines of MXML total.  But the ToggleButtonBar does not behave the same way as when I add it to the TDJ project, so I know I am missing something related to CSS configuration.  I am hoping that looking at this small project, you might quickly see it as I have lost three days on this.  Here is the zip of the project.


[CSSTest.zip](https://github.com/apache/royale-asjs/files/5963290/CSSTest.zip)

**Comment by carlosrovira** [comment]
Hi,

no issues with CSS, I found a NPE in ToggleButtonBarView that I just committed. This NPE can be seen in the browser console. It's important you take a look at the console to catch this kind of problems since are of great help.

Also I saw the actual way to setup the data provider is making the component not work as expected. I need to investigate further and fix the selected index issue (-1) since the latest changes still are not completed.

To solve the problem, please update to my last commit and then change the `spinner.mxml` to this code. That should make all work right:

```xml
<j:SectionContent xmlns:fx="http://ns.adobe.com/mxml/2009"
        xmlns:j="library://ns.apache.org/royale/jewel"
        xmlns:js="library://ns.apache.org/royale/basic"
        xmlns:html="library://ns.apache.org/royale/html"
        xmlns:c="components.*">

    <j:Group localId="spinnerGroup" width="100%" height="100%">
        <j:beads>
            <j:VerticalCenteredLayout/>
        </j:beads>

        <j:ToggleButtonBar localId="toggleButtonBar" emphasis="primary" >
            <j:dataProvider>
                <js:ArrayList source="[Sale, Rental]" />
            </j:dataProvider>
        </j:ToggleButtonBar>

    </j:Group>

</j:SectionContent>
```

**Comment by carlosrovira** [comment]
Hi @Bilbosax, please take a look at the changes I just committed. I think all should be now ok for ToggleButtonBar.

**Comment by carlosrovira** [comment]
Hi @Bilbosax,

I get the time to try adding dataProvider at initComplete and is now working too. I added to TDJ in my latest commit.
Altough I recommend you to use the inline mxml I showed you above since is a better less verbose approach.

So I consider ToggleButtonBar is finished with this changes. Let me know if you find it completed too.

**Comment by Bilbosax** [comment]
On your previous commit, it was still not working.  I could get it to work on a small sample project, but once the project got big, it stopped performing correctly  It was if something important wasn't getting initialized in time.  I will have to get back with you as I cannot install the latest nightly Royale SDK using Moonshine.  I am getting this error:

Installer path: C:\Program Files\MoonshineSDKInstaller
Installer version 3.7.0 (windows)
Available Memory: 39956480 (64-bit)
Using Locale: en_US
Fetched the SDK download mirror URL from the CGI.
Downloading royale from: http://apacheroyaleci2.westus2.cloudapp.azure.com:8080/job/royale-asjs_jsonly/lastSuccessfulBuild/artifact/out/apache-royale-0.9.8-bin-js.zip
Unable to download royale


I will go over to their Github page an inquire, and then I will get back with you on the ToggleButtonBar

**Comment by carlosrovira** [comment]
@aharui it seems nightly server is down can you take a look?

**Comment by carlosrovira** [comment]
@Bilbosax, until links are on again, I just can suggest to build from sources. I think it could be a bit cumbersome at first...but it's the best to take control over Royale instead of depending on this kind of links. But it's just a suggestion.

**Comment by carlosrovira** [comment]
Hi @Bilbosax, nightly URLs seems to work again: https://royale.apache.org/get-involved/

**Comment by Bilbosax** [comment]
Hi Carlos.  The server is back up and I am compiling with your latest commits.  Sadly, there is still an issue with the ToggleButtonBar.  I have attached a VS Code example at the bottom for you to take a look at.  The component seems to work in some situations, and not others. When you run it, you will see that under the current setup, the selectedIndex=0 is never initialized and no button is selected by default, and once a button is selected, it can be deselected by clicking it again.  You will find the ToggleButtonBar located in the search.mxml component.

[ToggleButtonBarIssue.zip](https://github.com/apache/royale-asjs/files/6004363/ToggleButtonBarIssue.zip)

**Comment by carlosrovira** [comment]
Hi Bilbo,
I just compiled and seems to work right for me. I'm using my SDK just built with ANT and VSCode + AS3&MXML to build your test with asconfigc file provided. I can see the first button selected, also I can't deselect it as expected.

I'll download nighlty and check with it now

**Comment by carlosrovira** [comment]
Hi Bilbo,

just downloaded nightly and is working fine too. Sorry but can't reproduce the problem since you project seems to work right.

**Comment by Bilbosax** [comment]
Man.  I feel so confused.  I have update Chrome and erased all history/cache.  I unistalled VS Code and reinstalled it.  I reinstalled AS3&MXML plugin.  I get no errors when I compile and my regular app logic all seems to work, but this button is not behaving on my machine as it is on yours in a project that I built.  Now that the Royale server is back up, I have downloaded the latest build that Moonshine will deliver. The only possible thing I can think of is that we are not compiling against the same thing for some reason.  Any thoughts at all on what else I should consider?

**Comment by carlosrovira** [comment]
Hi Bilbo,

to be sure, please check this link: http://apacheroyaleci2.westus2.cloudapp.azure.com:8080/job/royale-asjs_jsonly/lastSuccessfulBuild/artifact/out/

then lets go to download the .zip bin file for example, this is: http://apacheroyaleci2.westus2.cloudapp.azure.com:8080/job/royale-asjs_jsonly/lastSuccessfulBuild/artifact/out/apache-royale-0.9.8-bin-js.zip

Unpack and then point VSCode to use this SDK and hit CTRL+ENTER

Let me know if this still fails

**Comment by Bilbosax** [comment]
OK. Using this latest download, I can confirm that the sample project I sent you compiles and behaves correctly.  But my actual app does not yet.  I am going to spend one more morning pealing off layers of code trying to isolate what makes it work or break and I will send you another zipped project to take a look at where it does not work.  I still think the component just works "some of the time", but under certain circumstances, it breaks.  Just haven't figured out where yet.

**Comment by Bilbosax** [comment]
Carlos, I have great news!  The ToggleButtonBar appears to work correctly as far as I can tell, in ALL situations.  But I am at a loss and I need you to educate me.

This is my first web app, so I am sure that there are things going on in the browser as well as the compiler that I am completely unaware of.  Instead of starting with my original project and tearing it down to find the error, I started with the zipped project I attached earlier that was compiling correctly with the latest build and started adding to it from the original project.  I copied one file at a time from my original project to the one that was compiling correctly, and by the time I was done, I had copied the entire original project over to the new project and it was working like a charm.  Two identical projects, yet one works, and one does not.  So I started thinking that perhaps it was a cache issue as you mentioned earlier in this thread.  I went to Google's website and read exactly how to clear the browser cache and followed the steps, but it did not work.  I still have a situation where I have two identical projects under different names, but one works and one does not.

**Comment by carlosrovira** [comment]
Hi @Bilbosax great you got it!! :D

About caching...Your problem is very strange, but for sure it will have an explanation, but is difficult to imagine what could be. I think we all have issues like that when start, even remember to have in Flex days with swfs.
What I do with browsers is to put my cache to 0 (or the lower number it allows) and if I still see something not updating right, push the delete cache button in the current browser (each one on in its own way). But I think currently I don't need to do nothing but say the browser not to store any file and force always to reload.

Another thing is when you put online, since you must to implement some way to ensure people load new content only when is new. For that there's some approaches like add a "?a=b" to a url, but thats covered in many internet pages.

You must find your way depending on OS, browser, tools and the way you build, but its normal to have issues when start with a tech and you're creating your development workflow

**Comment by Bilbosax** [comment]
I edited my last answer a lot, so I don't know if you read the final version, but what I finally found the issue to be was the bin folder.  

"Finally, I thought that perhaps the compiler was not updating or overwritting the ToggleButtonBar in the bin directory correctly. I deleted the bin directory to let the compiler write to it fresh, and voila, it worked! So long story short, I had to delete the bin directory to make the ToggleButtonBar work. So my question is this: Is it common to need to delete the bin folder occasionally in web design, or is there perhaps a small glitch in the compiler that needs to be addressed regarding the ToggleButtonBar? I find it odd that I have no problems with any of the other components, but I do with the ToggleButtonBar."

Could there be a problem with the compiler not overwriting the ToggleButtonBar in the bin folder?

**Comment by carlosrovira** [comment]
Hi @Bilbosax, can't say really since I don't use that way to compile. I build with maven alway to target folder since that ensure my build works the same in local and in a server (or continuous integration system). Maybe @joshtynjala can let us know about the need to clean the bin folder with VSCode and his extension, maybe there's some flag or something.

**Comment by joshtynjala** [comment]
Any time that you update the Royale SDK to a new version, you need to clean your project's output directory. As I understand it, the compiler has some kind of optimization where it won't replace a .js file in the output directory if it already exists because that makes builds complete faster.

**Comment by Bilbosax** [comment]
Thanks Josh.  Appreciate you dropping by and giving us your input.
